File: src/handlers/xdg_decoration.rs
use smithay::wayland::shell::xdg::{
    decoration::XdgDecorationHandler,
    ToplevelSurface,
};
use wayland_protocols::xdg::decoration::zv1::server::zxdg_toplevel_decoration_v1::Mode;

use crate::state::PancakeState;

impl XdgDecorationHandler for PancakeState {
    fn new_decoration(&mut self, toplevel: ToplevelSurface) {
        // Tell every client to use client-side decorations — Pancake doesn't
        // draw server-side title bars yet. Apps that honour this (GTK4, Qt6)
        // will draw their own CSD; ones that ignore it behave as before.
        toplevel.with_pending_state(|state| {
            state.decoration_mode = Some(Mode::ClientSide);
        });
        toplevel.send_pending_configure();
    }

    fn request_mode(&mut self, toplevel: ToplevelSurface, mode: Mode) {
        let granted = match mode {
            Mode::ServerSide => Mode::ClientSide,
            _ => mode,
        };
        toplevel.with_pending_state(|state| {
            state.decoration_mode = Some(granted);
        });
        toplevel.send_pending_configure();
    }

    fn unset_mode(&mut self, toplevel: ToplevelSurface) {
        toplevel.with_pending_state(|state| {
            state.decoration_mode = None;
        });
        toplevel.send_pending_configure();
    }

}

File: csharp/Pancake.Wayland/XdgDecoration.cs
using NWayland;
using NWayland.Protocols.XdgDecorationUnstableV1;

namespace Pancake.Wayland;

// Full port of src/handlers/xdg_decoration.rs. Unlike compositor.rs/
// xdg_shell.rs, this handler has no dependency on Space/Window at all --
// it's pure protocol-level mode negotiation (tell clients whether to draw
// their own title bar or let the compositor draw one), so this is a
// complete, not partial, port.
internal sealed class ZxdgDecorationManagerListener : ZxdgDecorationManagerV1.ServerListener
{
    protected override void Destroy(ZxdgDecorationManagerV1.Server resource) => resource.Dispose();

    protected override void GetToplevelDecoration(ZxdgDecorationManagerV1.Server resource,
        NewId<ZxdgToplevelDecorationV1.Server, ZxdgToplevelDecorationV1.ServerListener> id,
        NWayland.Protocols.XdgShell.XdgToplevel.Server? toplevel)
    {
        var decoration = id.GetAndConsume(new ZxdgToplevelDecorationListener());
        // Port of new_decoration: Pancake doesn't draw server-side title
        // bars yet, so always start every toplevel as client-side-decorated.
        decoration.Configure(ZxdgToplevelDecorationV1.ModeEnum.ClientSide);
    }
}

internal sealed class ZxdgToplevelDecorationListener : ZxdgToplevelDecorationV1.ServerListener
{
    protected override void Destroy(ZxdgToplevelDecorationV1.Server resource) => resource.Dispose();

    // Port of request_mode: ServerSide requests are downgraded to
    // ClientSide (same "doesn't draw SSD yet" reason as above); any other
    // requested mode is granted as-is.
    protected override void SetMode(ZxdgToplevelDecorationV1.Server resource, ZxdgToplevelDecorationV1.ModeEnum mode)
    {
        var granted = mode == ZxdgToplevelDecorationV1.ModeEnum.ServerSide
            ? ZxdgToplevelDecorationV1.ModeEnum.ClientSide
            : mode;
        resource.Configure(granted);
    }

    // Port of unset_mode: falls back to the compositor's default choice,
    // same as a freshly created decoration object.
    protected override void UnsetMode(ZxdgToplevelDecorationV1.Server resource) =>
        resource.Configure(ZxdgToplevelDecorationV1.ModeEnum.ClientSide);
}
